### PR TITLE
adding `flymake` to `:checkers syntax` 

### PR DESCRIPTION
**Pull request by jeetelongname**
There is more to come, most of it finding checkers, but there should be a point
(very soon) where if you use lsp with all of your langs this is a drop in
replacement.
I need to also write an alternative too `+emacs-lisp-reduce-flycheck-errors-in-emacs-config-h` but that shall come in due course 

-----
- [X] I searched the issue tracker and this hasn't been PRed before.
- [X] My changes are not on [the do-not-PR list](https://doomemacs.org/d/do-not-pr) for this project.
- [X] My commits conform to [the git conventions](https://doomemacs.org/d/git-conventions).
- [ ] My changes are visual; I've included before and after screenshots.
- [ ] I am blindly checking these off.
- [X] Any relevant issues or PRs have been linked to.
- [X] This a draft PR; I need more time to finish it.

### CODE REVIEW

File: modules/checkers/syntax/config.el [hunk 1]
```diff
@@ -81,3 +84,15 @@
 
 ;;
 ;;; TODO Flymake
```
Comment (gagbo) [inline modules/checkers/syntax/config.el:86]: Might as well remove the TODO item

File: modules/checkers/syntax/config.el [hunk 2]
```diff
@@ -81,3 +84,15 @@
...
+
+(use-package flymake
+  :when (featurep! +flymake)
+  :defer t
+  :init
+  (add-hook! (prog-mode text-mode) #'flymake-mode)
+  :config
+  (setq flymake-fringe-indicator-position 'right-fringe))
+
+(use-package flymake-popon
```
Comment (gagbo) [inline modules/checkers/syntax/config.el:96]: Do we want to hide that behind the childframe feature ?

File: modules/checkers/syntax/packages.el [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
-(package! flycheck-popup-tip :pin "ef86aad907f27ca076859d8d9416f4f7727619c6")
-(when (featurep! +childframe)
-  (package! flycheck-posframe :pin "8f60c9bf124ab9597d681504a73fdf116a0bde12"))
+(unless (featurep! +flymake)
+  (package! flycheck :pin "784f184cdd9f9cb4e3dbb997c09d93e954142842")
+  (package! flycheck-popup-tip :pin "ef86aad907f27ca076859d8d9416f4f7727619c6")
+  (when (featurep! +childframe)
+    (package! flycheck-posframe :pin "8f60c9bf124ab9597d681504a73fdf116a0bde12")))
 
 ;; TODO flymake?
```
Comment (gagbo) [inline modules/checkers/syntax/packages.el:10]: Might as well remove the TODO item

**Review by gagbo — COMMENTED** [review]
Small nitpicks and general warning: I know that when I touched lsp-mode and eglot checkers I always worked under the assumption that `:checkers syntax` is equivalent to using flycheck (necessary and sufficient). So you might need to make a pass there to make sure that flycheck stuff is correctly inactive when `(featurep! :checkers syntax +flymake)`

**Comment by jeetelongname** [comment]
I wanted to get the stub up but I am aware of that, I would need to do a pass over a lot of the modules to make them more generic. But that is on the agenda.

File: modules/checkers/syntax/config.el (hunk 2 above)
Comment (jeetelongname) [reply modules/checkers/syntax/config.el:96]: Thats the thing. Its not using childframes. I guess it depends on the spirit of what `+childframe` means.

Comment (jeetelongname) [reply modules/checkers/syntax/config.el:96]: Nevermind I just feature gated it properly, postframes will be used when `+childframe` is enabled

File: modules/checkers/syntax/packages.el [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
+  (package! flycheck :pin "784f184cdd9f9cb4e3dbb997c09d93e954142842")
+  (package! flycheck-popup-tip :pin "ef86aad907f27ca076859d8d9416f4f7727619c6")
+  (when (modulep! +childframe)
+    (package! flycheck-posframe :pin "8f60c9bf124ab9597d681504a73fdf116a0bde12")))
 
-;; TODO flymake?
+;; Flymake
+(when (modulep! +flymake)
+  (package! popon :recipe (:repo "https://codeberg.org/akib/emacs-popon"))
+  (package! flymake-popon :recipe (:repo "https://codeberg.org/akib/emacs-flymake-popon")))
```
Comment (gagbo) [inline modules/checkers/syntax/packages.el:13]: With straight.el having added support for NonGNU ELPA you shouldn't need to specify the recipes anymore. Is current straight.el pin too old ?

Comment (jeetelongname) [reply modules/checkers/syntax/packages.el:13]: I think so, I am getting an error still, I can add a NOTE to it though that I can remove the recipe when its bumped

**Comment by peterhoeg** [comment]
This might be slightly off-topic but how does flymake compare to flycheck?

**Comment by jeetelongname** [comment]
- flymake is built in first of all. Thats its biggest advantage, 
- It also does not need as much configuration as well with its defaults 
- It composes well with emacs systems (with diagnostic functions just being added to a hook as one example) 
- From what i am seeing, its the way the community is going.

atm both are usable, I have been using this branch for a while now (other than a small break for a large bump of the tree sitter module) there are some loose ends, I need to replace a lot of the flycheckers I disabled but other than that, especially if you use lsp for most things it just works:tm:

**Comment by peterhoeg** [comment]
`eglot` is still pulling in `flycheck-eglot`

I'm not sure what the official position is on narrowly used helpers, but I end up with a fair bit of `(and (modulep! :checkers syntax) (not (modulep! :checkers syntax +flymake)))`. It would be very nice to have a helper defined to avoid that.

File: modules/lang/cc/config.el [hunk 5]
```diff
@@ -52,20 +52,20 @@ This is ignored by ccls.")
...
+                  :true "true" :false "false"
+                  :int "int" :float "float"
+                  :str "std::string"
+                  :bool "bool"
+                  ;; Flow
+                  :not "!"
+                  :and "&&" :or "||"
+                  :for "for"
+                  :return "return"
+                  :yield "#require")
```
Comment (hlissner) [inline modules/lang/cc/config.el:68]: Please revert

File: modules/lang/clojure/config.el [hunk 6]
```diff
@@ -171,85 +171,85 @@
...
+          "r" #'cider-ns-refresh
+          "R" #'cider-restart
+          "c" #'cider-repl-clear-buffer)
+         :map cider-repl-history-mode-map
+         :i [return]  #'cider-repl-history-insert-and-quit
+         :i "q"  #'cider-repl-history-quit
+         :i "l"  #'cider-repl-history-occur
+         :i "s"  #'cider-repl-history-search-forward
+         :i "r"  #'cider-repl-history-search-backward
+         :i "U"  #'cider-repl-history-undo-other-window)))
```
Comment (hlissner) [inline modules/lang/clojure/config.el:252]: Please revert

File: modules/lang/clojure/packages.el [hunk 7]
```diff
@@ -14,10 +14,11 @@
 (package! parseedn :pin "a09686fbb9113b8b1b4f20c9e1dc0d6fea01a64f")
 
 ;;; Core packages
-(package! clojure-mode :pin "3453cd229b412227aaffd1dc2870fa8fa213c5b1")
-(package! clj-refactor :pin "b5abe655e572a6ecfed02bb8164b64716ef76b8e")
-(package! cider :pin "1ed5163433c991c00ea83fdd4447e8daf4aeccbe")
-(when (modulep! :checkers syntax)
+(package! clojure-mode :pin "414157c3e523e80cc44dca8f86f1853122ee5f6b")
+(package! clj-refactor :pin "db89b558710d31d4b9dfe7c4d2a07df9b498d4a6")
+(package! cider :pin "92c1b6de20805a9caa556cc671ac39080e4e89cc")
```
Comment (hlissner) [inline modules/lang/clojure/packages.el:19]: Please revert (or are specific versions needed for flymake support?)

File: modules/tools/lsp/+eglot.el [hunk 8]
```diff
@@ -25,6 +26,11 @@
 
   (add-to-list 'doom-debug-variables '(eglot-events-buffer-size . 0))
 
+  (when (and (modulep! :checkers syntax)
+             (not (modulep! :checkers syntax +flymake)))
+    (after! flycheck
+      (load! "autoload/flycheck-eglot")))
+
```
Comment (hlissner) [inline modules/tools/lsp/+eglot.el:33]: This seems outdated with https://github.com/doomemacs/doomemacs/blob/master/modules/tools/lsp/+eglot.el#L64-L66

**Review by hlissner — CHANGES_REQUESTED** [review]
Is there something holding up this PR?

> It would be very nice to have a helper defined to avoid that.

Will do. In v3, `modulep!` will understand negated flags as an implicit `(not ...)`, e.g. `(modulep! :checkers syntax -flymake)` => returns `t` only if the module is enabled _and_ lacks `+flymake`.

**Comment by jeetelongname** [comment]
I wanted to add flymake support for a few more modules but everything works if you use lsp (like I do). Ill see if I can revert these changes within the next couple of days

File: modules/tools/lsp/+eglot.el (hunk 8 above)
Comment (jeetelongname) [reply modules/tools/lsp/+eglot.el:41]: Looks like it. Ill remove that and properly feat gate flycheck eglot in that case

**Comment by jeetelongname** [comment]
that should be all done. if there is anything more let me know.

**Comment by jeetelongname** [comment]
Also I have not completed this. 
> I need to also write an alternative too +emacs-lisp-reduce-flycheck-errors-in-emacs-config-h but that shall come in due course

I may be able to have a go at it over the next couple of days but feel free to merge this PR and or tackle this yourself.

**Review by hlissner — APPROVED** [review]

**Comment by hlissner** [comment]
LGTM! Thanks again for your help!

> I may be able to have a go at it over the next couple of days but feel free to merge this PR and or tackle this yourself.

Will do!

**Comment by artpelling** [comment]
I can't find this feature in the docs. Is it fully functional with `eglot`? I have some issues getting `flymake` plugins to work.

**Comment by hlissner** [comment]
@artpelling https://github.com/doomemacs/doomemacs/blob/master/modules/checkers/syntax/README.org#module-flags

Works fine with Eglot. Please open a new issue (or [discussion](https://doomemacs.org/discuss) if you're uncertain about the cause).

**Pull request merged**